ui/src/app/lib/atlasmap-data-mapper/components/mapping/mapping.list.component.ts

```diff
@@ -127,8 +127,10 @@ export class MappingListFieldComponent {
                                     <div class="clear"></div>
                                 </div>
                                 <div class="error">
-                                    <span class="pficon pficon-error-circle-o"
-                                    *ngIf="mapping.validationErrors.length"></span>
+                                    <span class="pficon"
+                                          [ngClass]="{'pficon-error-circle-o': mapping.hasValidationErrors(),
+                                                      'pficon-warning-triangle-o': mapping.hasValidationWarnings()}"
+                                          *ngIf="mapping.validationErrors.length"></span>
```

Comment:
On every cycle of Angular's ChangeDetector mechanism, you'll be re-executing `mapping.hasValidationErrors()` and `mapping.hasValidationWarnings()`, which will execute subsequent functions inside the `MappingModel` class such as `getValidationErrors ()` and `getValidationWarnings()` which will also execute the `this.validationErrors.filter` functional lambda and the like... These are too many executions happening on every tick. Now multiply this for an application with literally hundreds of components (which AtlasMapper might become one day) and you'll bump into a severe performance bottleneck down the track.

Let me suggest a more convenient and Angular-friendly approach for this: Use a pure`@Pipe` instead. Something like this: 

```
<span class="pficon" [class.pficon-error-circle-o]="mapping.validationErrors | hasErrors" ...> 
``` 

Where `hasErrors` is a pipe that will receive `mapping.validationErrors` and will inspect if there is any actual error by returning `mapping.validationErrors.some(e => e.level >= ErrorLevel.ERROR)`.

The only caveat is that you'll want to alter the memory reference for `mapping.validationErrors` every time you update it, so it becomes immutable:

```
addValidationError(message: string) {
  const e = new ErrorInfo(message, ErrorLevel.VALIDATION_ERROR);
  this.validationErrors = [... this.validationErrors, e];
}
``` 

BTW, since `public` is implied in TypeScript entities, the industry convention is to never annotate class members with it   (which unfortunately we do not honor here)

Reply:
Does your suggestion reduce the number of times we evaluate the `validationErrors` array or does it just reduce the method calls? To improve performance, wouldn't be better to create an array of view objects with only the data we need to show on that table? We would have a simple string field with the class name already set on each object.

Reply:
My proposal aims to minimize function calls, which subsequently will reduce the amount of times we evaluate `validationErrors` under the hood. 

Taking this approach will make our application operate all the way around. The template and its component controller will not evaluate `MappingModel.getValidationErrors()` anymore. On the contrary they will evaluate just nothing after content is checked. In the very moment that the ChangeDetection system detects that `MappingModel.validationErrors` has changed its type reference, an OnPush operation will be triggered throughout the component tree and will force the pipe to return a new value.

Encapsulating this conditional logic in a pipe also has an additional advantage: we can reuse this  logic elsewhere at zero cost.

Reply:
OK, got it. By referencing `MappingModel.validationErrors` directly instead of a method, Angular can detect when it changes and evaluate it only at that point, correct?

Reply:
>By referencing `MappingModel.validationErrors` directly instead of a method, Angular can detect when it changes

Yes, but here I'm proposing to funnel the conditional validation through a pipe, and this requires taking something into consideration: Unless defined as IMPURE, pipes are only affected by the ChangeDetector system when the input argument (`validationErrors`) changes its type reference. This is perfectly fine when such input is a primitive type (strings, numbers) but `validationErrors` is an array, and adding or removing items will not change its type reference. It will become the same instance object and therefore ChangeDetection will not happen. That's why I pointed out to create a new object by means of ES6's spread operator.